### ISSUE TITLE
CSV output too many newlines

### ISSUE THREAD

**Issue by JanDeneweth-bmx** [sourmash-bio/sourmash#1318]
We noticed on using sourmash with option `--csv`, the output contains 2 instead of 1 newline between each line. This is the result of opening the files incorrectly. Files must be opened with argument `newline=''` (https://docs.python.org/3.8/library/csv.html?highlight=csv#csv.writer).

I hope this is the full extent of the bug and can be resolved easily.

**Comment by ctb** [comment]
thanks for reporting this, @JanDeneweth-bmx! Which command(s) does this affect, and what is the practical impact that caused you to notice it, so we can properly test it? (I've been using both spreadsheet programs and csv.reader/DictReader with the sourmash output and haven't had any problems, so I'm not quite sure how to test for this!)

**Comment by JanDeneweth-bmx** [comment]
Hi ctb, thanks for the quick response!

We detected this using the `--csv` output option for the `compare` command. The output contained extra empty newlines between each line, making the file incompatible with standard CSV parsers (we need to feed to output into a non-python program). 

After quickly browsing all uses of the csv library in the code it would seem all usage (writers and readers) on all places is incorrect. In some places fixing may be slightly more work because you have custom 'FileOutput' classes which wrap opening the files, and do not yet except a 'newline' parameter. The class needs to be changed as well.

**File -> Functions**
sourmash/command.py -> compare, plot, import_csv, search, categorize (twice), gather, multigather
sourmash/lca/command_classify.py ->classify (twice)
sourmash/lca/command_gather -> gather_main
sourmash/lca/index ->load_taxonomy_assignments
sourmash/lca/command_summarize.py -> output_csv (possibly, the filehandle is created elsewhere)
sourmash/sig/__main__ -> describe

Regrettably, this is on company time and I can't spend too much time on it, otherwise I'd have set up a pull request with fixes.

**Comment by ctb** [comment]
ok, thanks! I was guessing it was "everywhere"! and no worries, it should be straightforward to fix.

**Comment by ctb** [comment]
There seems to be no operational distinction on either Mac OS X or Linux to setting newline -- see #1319 for code. I might be missing something, will recheck in a bit!

(see https://hackmd.io/iiZdj22YT0evJcAbvdAhwA for a colored syntax highlighting of the below.)

below, `old.csv` is with `latest` branch that does not correctly set newline, `new.csv` is with `fix/csv_newline` branch that _does_ set newline.

On Mac OS X, python 3.7.6:
```python=
>>> data1 = open('old.csv', 'rb').read()
>>> data1
b'GCF_000005845.2_ASM584v2_genomic.fna\r\n1.0\r\n'
>>> data2 = open('old.csv', 'rt').read()
>>> data2
'GCF_000005845.2_ASM584v2_genomic.fna\n1.0\n'
>>> data3 = open('new.csv', 'rb').read()
>>> data4 = open('new.csv', 'rt').read()
>>> data3
b'GCF_000005845.2_ASM584v2_genomic.fna\r\n1.0\r\n'
>>> data4
'GCF_000005845.2_ASM584v2_genomic.fna\n1.0\n'
>>> 
```

On Linux, python 3.7.9:

```python=
>>> data1 = open('old.csv', 'rb').read()
>>> data1
b'GCF_000005845.2_ASM584v2_genomic.fna.gz\r\n1.0\r\n'
>>> data2 = open('old.csv', 'rt').read()
>>> data2
'GCF_000005845.2_ASM584v2_genomic.fna.gz\n1.0\n'
>>> data3 = open('new.csv', 'rb').read()
>>> data4 = open('new.csv', 'rt').read()
>>> data3
b'GCF_000005845.2_ASM584v2_genomic.fna.gz\r\n1.0\r\n'
>>> data4
'GCF_000005845.2_ASM584v2_genomic.fna.gz\n1.0\n'
```

**Comment by ctb** [comment]
note, even if the above holds true on Mac OS X and Linux, there may be a distinction on Windows! We just don't support windows for sourmash.

**Comment by JanDeneweth-bmx** [comment]
Oh, that probably explains it! I should've clarified, we are using Windows. Thanks to Rust and Python being mostly OS agnostic, it actually works very well! We didn't have to change the code if I remember correctly, and haven't encountered any real issues yet. The CSV output is a minor annoyance. This is in stark contrast to Mash, which we managed to get working on Windows with much effort and some changes to the code, but for unclear reasons generates different result when comparing with Linux.

If you want to close this issue because Windows isn't supported anyway, and linux doesn't suffer from the issue (despite what CSV lib claims), I can understand. Just know that Windows support is apparently quite feasable should you wish, and an official Windows-compatible minhashing toolset does not yet exist as far as I know. 

If you are interested in our experiences using sourmash on Windows, let me know. I don't think I can disclose the exact product this development is linked to yet, but we develop a desktop application that - amongst other things - integrates many Bioinformatics analyses, with many large users for over 15 years. Might be useful feedback should you consider Windows support.

**Comment by JanDeneweth-bmx** [comment]
I really apologize for not mentioning it earlier. Since it was an issue in the python rather than the Rust I didn't consider there may have been an OS difference. But should'be known to at least mention it since we are using the program outside of the systems it supports.

**Comment by ctb** [comment]
no worries at all! I think it is valuable for forwards compatibility to do the right thing for CSV files, and we are happy to be alerted to the problem!

I am actually fascinated to hear that sourmash works on Windows! 😆

**Comment by luizirber** [comment]
> Oh, that probably explains it! I should've clarified, we are using Windows. Thanks to Rust and Python being mostly OS agnostic, it actually works very well! We didn't have to change the code if I remember correctly, and haven't encountered any real issues yet. 

We are not very far from actually building wheels for Windows (we already build packages for windows in [conda-forge][0]). The last hurdle when I set it up was khmer, but we don't depend on it anymore for `Nodegraph`s, so should be straightforward to add.
Are you building with the mingw or msvc toolchain?

[0]: https://anaconda.org/conda-forge/sourmash-minimal

> If you want to close this issue because Windows isn't supported anyway, and linux doesn't suffer from the issue (despite what CSV lib claims), I can understand. Just know that Windows support is apparently quite feasable should you wish, and an official Windows-compatible minhashing toolset does not yet exist as far as I know.

We can add another check on CI for running the tests on windows too.

> If you are interested in our experiences using sourmash on Windows, let me know. I don't think I can disclose the exact product this development is linked to yet, but we develop a desktop application that - amongst other things - integrates many Bioinformatics analyses, with many large users for over 15 years. Might be useful feedback should you consider Windows support.

I very much am interested, so if this convo goes into an email (in case you want to avoid disclosing info in public) please [add me](mailto:lcirberjr@ucdavis.edu) =]

**Comment by ctb** [comment]
> > If you are interested in our experiences using sourmash on Windows, let me know. I don't think I can disclose the exact product this development is linked to yet, but we develop a desktop application that - amongst other things - integrates many Bioinformatics analyses, with many large users for over 15 years. Might be useful feedback should you consider Windows support.
> 
> I very much am interested, so if this convo goes into an email (in case you want to avoid disclosing info in public) please [add me](mailto:lcirberjr@ucdavis.edu) =]

me too, titus@idyll.org!

Regardless of specific info or non-disclosure issues, we would be happy to discuss how we can better support your use cases by adding tests to our continuous integration. My motivation here is that you are likely encountering problems that other potential users are facing, and it would definitely be good to both fix them and get those added to our set of regular checks!

**Comment by ctb** [comment]
hello @JanDeneweth-bmx, we just merged a fix for this issue, over in #1319.

https://github.com/dib-lab/sourmash/pull/1334 is adding wheel building for Windows, and I've just updated that from the `latest` branch that now contains #1319, so I'll monitor those tests. Once wheel building for Windows works, you can give the new CSV output a try! (Or of course you can build & install from `latest` if you like ;).

This fix will be released in the next version of sourmash, along with other recent changes. We don't have a firm date for that release but it will be sourmash v4.1.0, probably by end of April.

**Comment by jandeneweth** [comment]
Thanks @ctb! We're currently too close to release to consider using a newer version right now, but this adds one more thing we look forward to in the coming update! Thanks again for the excellent work :)
